src/Pirouette/Symbolic/Eval.hs

```diff
@@ -311,12 +323,30 @@ symEvalOneStep t@(R.App hd args) = case hd of
     mayBranches <- lift $ branchesBuiltinTerm @lang builtin translator args
     case mayBranches of
       -- if successful, open all the branches
-      Just branches -> asum $
-        flip map branches $ \(SMT.Branch additionalInfo newTerm) -> do
-          lift $ learn additionalInfo
-          consumeFuel
-          signalEvaluation
-          pure newTerm
+      Just branches -> do
+        -- do
+        -- consumeFuel
+        -- signalEvaluation
+        -- lift $ learnBranches branches
+        --
+        asum $
+          flip map branches $ \(SMT.Branch additionalInfo newTerm) -> do
+            lift $ learn additionalInfo
+            consumeFuel
+            signalEvaluation
+            pure newTerm
```

Comment:
This is how branching is handled inside Pirouette. For each branch, we add the corresponding constraints to each possible state (gathered in the `WeightedList` inside the `SymEval` monad) and prune the (current state + current branch) if the constraints are inconsistent with the state. Asking the solver to check for this consistency is done in parallel (i.e., up to one solver for each possible current state). Finally, once we've done this for every branch (sequentially), we gather all the possible new states with `asum`. In a nutshell, the scheduling can be described as
```
solve st1 cst1 || solve st1 cst2 || ...;
solve st2 cst1 || solve st2 cst2 || ...;
...
```